rfcs/90-parse-pairs-from-string.md

```diff
@@ -0,0 +1,106 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-06-01
+deprecates:
+  - 13-remove-config-string.md
+---
+
+# GSP-90: Parse Pairs from String
+
+## Background
+
+We had config strings like `s3://bucket_name?access_key=xxxx&secret_key=xxxx` before. This is introduced in [GSP-3] but deprecated by [GSP-13].
+
+There are two reasons why config strings are removed, as mentioned in [GSP-13]:
+
+1. Usability: If an application uses [go-storage]'s config string, then it either exposes the format of config string to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers). And if the application wants to combine its other config into a config string, this is also inconvenient.
+2. Implementation: We didn't find a good way to parse pairs from strings easily and safely.
+
+So we let the users of [go-storage] construct pairs directly using `WithXxx()`. 
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose the following mechanism for parsing pairs:
+
+### Pair Registry
+
+We first register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+type PairMap map[string]reflect.Type
+
+func RegisterServicePairMap(ty string, m PairMap) {}
+```
+
+From services side, we can generate following code:
+
+```go
+func servicePairs() PairMap {}
+
+func init() {
+	// ...
+	pairs.RegisterServicePairMap("<type>", servicePairs())
+}
+```
+
+The pair information maps are also generated.
+
+### Parse According to Type
+
+For each type, we can have a parse function.
+
+```go 
+type parseFunc func(string) (interface{}, error)
+
+var parseMap map[reflect.Type]parseFunc
+```
+
+Then we can parse a pair of string into a pair. We check the pair's name in a `PairMap` and get its type information, then use corresponding parse function to parse the pair's value.
+
+```go
+func (m PairMap) parse(k string, v string) (pair Pair, err error) {}
+```
+
+Then we can easily provide APIs for users to parse pairs:
+
+```go 
+func Parse(ty string, k string, v string) (pair Pair, err error) {}
```

Comment:
I think the API exposed to users is `pairs.FromString(typ string, v string) (ps []Pair, err error)`.

And we don't need to support parse single pair: we only allow users to use these functions to do initialization.

Reply:
> we only allow users to use these functions to do initialization

You mean only support something like `Open(config string) Storager`, instead of `Parse(config string) []Pair`? 
If so, users cannot add other pairs into config string, as you mentioned earlier.
If not, what's the difference? (btw I agree that parsing single pair API is too fine-grained)

Reply:
For now, we have APIs like `s3.NewStorager(ps ..Pair) (Storager, error)` and we build a new API in service registry like `NewStorager(typ string, ps ...Pair) (Storager, error)`.

In this GSP, we will add functions that can: parse from string to type and pairs so that users can use them in this way:

```go
typ, ps, err := FromString("s3://xxxxx@yyyyy")
if err != nil {
    ...
}
store, err := NewStorager(typ, ps...)
```

> API should be defined in this GSP, and `FromString` is just an example.

And `FromString` is different from `FromMap` (please ignore the API I mentioned in above comment):

- `FromString` doesn't need an extra type field, we can add it into our format.
- `FromMap` needs an extra type field, like `FromMap(typ string, ps ...Pair)`

Reply:
Then there's no misunderstanding. But this does not meet
> we only allow users to use these functions to do initialization.